ipa-core/src/report/hybrid.rs

```diff
@@ -150,37 +151,94 @@ impl TryFrom<Bytes> for EncryptedHybridReport {
     }
 }
 
+const TAG_SIZE: usize = TagSize::USIZE;
+
+#[derive(Clone, Debug)]
+pub struct UniqueTag {
+    bytes: [u8; TAG_SIZE],
+}
+
 pub trait UniqueBytes {
-    fn unique_bytes(&self) -> Vec<u8>;
+    fn unique_bytes(&self) -> [u8; TAG_SIZE];
+}
+
+impl UniqueBytes for UniqueTag {
+    fn unique_bytes(&self) -> [u8; TAG_SIZE] {
+        self.bytes
+    }
 }
 
 impl UniqueBytes for EncryptedHybridReport {
     /// We use the `TagSize` (the first 16 bytes of the ciphertext) for collision-detection
     /// See [analysis here for uniqueness](https://eprint.iacr.org/2019/624)
-    fn unique_bytes(&self) -> Vec<u8> {
-        self.mk_ciphertext()[0..TagSize::USIZE].to_vec()
+    fn unique_bytes(&self) -> [u8; TAG_SIZE] {
+        let slice = &self.mk_ciphertext()[0..TAG_SIZE];
+        let mut array = [0u8; TAG_SIZE];
+        array.copy_from_slice(slice);
+        array
+    }
+}
+
+impl UniqueTag {
+    // Function to attempt to create a UniqueTag from a UniqueBytes implementor
+    pub fn from_unique_bytes<T: UniqueBytes>(item: &T) -> Self {
+        UniqueTag {
+            bytes: item.unique_bytes(),
+        }
+    }
+
+    /// Maps the tag into a consistent shard.
+    /// Note that `ShardIndex` is limited to u32, so we only use the first 4 bytes.
+    ///
+    /// ## Panics
+    /// if the `TAG_SIZE < 4`
+    /// note: ~10 below this, we have a compile time check that `TAG_SIZE = 16`
+    #[must_use]
+    pub fn shard_picker(&self, shard_count: ShardIndex) -> ShardIndex {
+        let num = u32::from_le_bytes(
+            self.bytes[0..4]
+                .try_into()
+                .expect("This is larger than 4 bytes"),
+        );
+        let shard_count = u32::from(shard_count);
+        ShardIndex::from(num % shard_count)
+    }
+}
+
+// This will vaild at compile time if TAG_SIZE doesn't match U16
+const _: [(); 16] = [(); TAG_SIZE];
```

Comment:
nice, you could use `const_assert!` macro for that to have even better error message

Reply:
or even `[(); U16::Size::USIZE]` could work

Reply:
oh cool, let me see if `const_assert!` works. basically I want something to fail if `TAG_SIZE` is no longer 16 bytes so we know to fix this.